### PR TITLE
ENT-638 Enterprise Offers

### PR DESCRIPTION
**Pull request by douglashall**
This PR adds the Enterprise offers feature. An Enterprise offer consists of a custom Condition model which checks that the user is linked to the specified EnterpriseCustomer, the course run(s) exist in that EnterpriseCustomer's catalogs, and the user has granted consent to share data for the course run(s).

Enterprise offers can be set up in the ecommerce administration tool.

ENT-638

**Testing Instructions:**
1. Visit [Enterprise Offers Admin](https://ecommerce-business.sandbox.edx.org/enterprise/offers) and verify that you can successfully create an Enterprise Offer using the existing [EnterpriseCustomerCatalogs](https://business.sandbox.edx.org/admin/enterprise/enterprisecustomercatalog/).
2. Visit https://business.sandbox.edx.org/enterprise/76afe510-a9fd-4b59-9ba1-49b537864205/course/course-v1:Microsoft+DAT206x+2017_T4/enroll/ as an unauthenticated user.
3. Log in through TestShib.
4. Verify that the course enrollment landing page displays the 50% Enterprise Offer discount.
5. Select Verified and click continue.
6. Verify the basket page displays the 50% Enterprise Offer discount.

### CODE REVIEW

File: ecommerce/enterprise/tests/mixins.py [hunk 1]
```diff
@@ -174,9 +174,12 @@ def mock_enterprise_learner_api(
...
                         {
-                            'user': 1,
-                            'state': 'enabled' if consent_provided else 'disabled',
-                            'enabled': consent_provided
+                            'username': 'verified',
+                            'enterprise_customer_uuid': enterprise_customer_uuid,
+                            'exists': consent_provided,
+                            'course_id': 'course-v1:FooX+FooX+Demo_Course',
+                            'consent_provided': consent_provided,
+                            'consent_required': consent_enabled
```
Comment (UmanShahzad) [inline ecommerce/enterprise/tests/mixins.py]: Ah, looks like I'm also doing this change over at https://github.com/edx/ecommerce/pull/1494/files, although I have some required differences -- up to you if you want to use what I have and put it in this PR or let that PR handle it, I'm neutral. Unless there's something I got wrong with my changes?

File: ecommerce/enterprise/offer/benefits.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+class EnterprisePercentageDiscountBenefit(BenefitWithoutRangeMixin, PercentageDiscountBenefit):
+    """ Enterprise-related PercentageDiscountBenefit without an attached range. """
+
+    class Meta(object):
+        app_label = 'enterprise'
+        proxy = True
+
+    @property
+    def name(self):
+        return _('{value}% enterprise discount').format(value=self.value)
```
Comment (UmanShahzad) [inline ecommerce/enterprise/offer/benefits.py]: I believe this can be safely removed because `BenefitWithoutRangeMixin`'s version would output the same thing. (Of course not true for the others).

File: ecommerce/enterprise/tests/mixins.py (hunk 1 above)
Comment (UmanShahzad) [reply ecommerce/enterprise/tests/mixins.py]: Additionally, mine also changes a doc in `ecommerce/enterprise/api.py`.

File: ecommerce/enterprise/offer/conditions.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        enterprise_customer = learner_data['enterprise_customer']
+        if self.enterprise_customer_uuid != enterprise_customer['uuid']:
+            # Learner is not linked to the EnterpriseCustomer associated with this condition.
+            return False
+
+        course_runs_with_consent = self._get_course_runs_with_consent(learner_data['data_sharing_consent_records'])
+        course_run_ids = []
+        for line in basket.all_lines():
+            product = line.product
```
Comment (UmanShahzad) [inline ecommerce/enterprise/offer/conditions.py]: I see that `product` isn't used elsewhere below, so maybe we can squish things into `course = line.product.course` to be succinct.

**Comment by UmanShahzad** [comment]
I like where this is headed and I think it should be enough back-end logic for the front-end work to come soon.

File: ecommerce/programs/conditions.py [hunk 4]
```diff
@@ -141,18 +142,3 @@ def get_applicable_lines(self, offer, basket, most_expensive_first=True):
...
-
-        Args:
-            offer (AbstractConditionalOffer)
-            basket (AbstractBasket)
-            affected_lines (tuple[]): The lines that have been affected by the discount.
-                This should be list of tuples (line, discount, qty)
-        """
-        for line, __, __ in affected_lines:
-            quantity_to_consume = min(line.quantity_without_discount, 1)
-            line.consume(quantity_to_consume)
```
Comment (UmanShahzad) [inline ecommerce/programs/conditions.py]: Just a nitty organizational suggestion: would it be worth moving `programs/condition.py` and `programs/benefits.py` into an `program/offer` folder, the way you've started for `enterprise/offer`? Or maybe since this existed beforehand, remove `enterprise/offer`? Following one way or the other across these apps helps a lot for seeing the landscape and similarities among them at a glance.

File: ecommerce/enterprise/offer/conditions.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                site.domain,
+                user.username,
+            )
+            return False
+        except IndexError:
+            # Learner is not linked to any EnterpriseCustomer.
+            return False
+
+        enterprise_customer = learner_data['enterprise_customer']
+        if self.enterprise_customer_uuid != enterprise_customer['uuid']:
```
Comment (UmanShahzad) [inline ecommerce/enterprise/offer/conditions.py]: We should change the left-hand side to `self.enterprise_customer['id']` since that's what our `ecommerce.enterprise.utils.get_enterprise_customer` function returns in the dictionary, and we presumably have access to `self.enterprise_customer` from the above `name` property.

File: ecommerce/enterprise/offer/conditions.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                # Basket contains products not related to a course_run.
+                return False
+
+            if enterprise_customer['enable_data_sharing_consent'] and course.id not in course_runs_with_consent:
+                # Basket contains course_runs for which the learner has not given consent to share data.
+                return False
+
+            course_run_ids.append(course.id)
+
+        if not catalog_contains_course_runs(basket.site, course_run_ids, self.enterprise_customer_uuid,
```
Comment (UmanShahzad) [inline ecommerce/enterprise/offer/conditions.py]: Same here as above with the enterprise customer UUID. But I think we can leave the catalog's UUID alone.

File: ecommerce/enterprise/offer/benefits.py (hunk 2 above)
Comment (douglashall) [reply ecommerce/enterprise/offer/benefits.py]: Good call.

File: ecommerce/enterprise/offer/conditions.py (hunk 3 above)
Comment (douglashall) [reply ecommerce/enterprise/offer/conditions.py]: Yup.

File: ecommerce/enterprise/tests/mixins.py (hunk 1 above)
Comment (douglashall) [reply ecommerce/enterprise/tests/mixins.py]: Yeah, we can work it out as we get closer to merging either this or that.

File: ecommerce/programs/conditions.py (hunk 4 above)
Comment (douglashall) [reply ecommerce/programs/conditions.py]: Yes, consistency is good. I kind of like the enterprise/offer directory structure, but I think it would be better to match what we already have for expediency's sake.

File: ecommerce/enterprise/offer/conditions.py (hunk 5 above)
Comment (douglashall) [reply ecommerce/enterprise/offer/conditions.py]: Actually, we don't have access to the enterprise_customer from the Condition object. I am going to denormalize and add both `enterprise_customer_uuid` and `enterprise_customer_name` fields to the Condition model since we won't be able to get the enterprise_customer_name using the enterprise_customer_uuid without the Site (in order to construct an API client) and we don't have access to the Site here either.

Comment (UmanShahzad) [reply ecommerce/enterprise/offer/conditions.py]: Yeah, you're right. That's why tests are failing on #1497 -- I tried putting `enterprise_customer` on the condition factory :)

File: ecommerce/enterprise/offer/conditions.py (hunk 6 above)
Comment (UmanShahzad) [reply ecommerce/enterprise/offer/conditions.py]: Nevermind.

File: ecommerce/enterprise/offer/conditions.py (hunk 5 above)
Comment (UmanShahzad) [reply ecommerce/enterprise/offer/conditions.py]: With that sorted out we should be well on our way, I believe, minus testing / any outstanding review for this overall work.

File: ecommerce/enterprise/forms.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            if hasattr(self.instance, 'condition'):
+                self.instance.condition.enterprise_customer_uuid = enterprise_customer_uuid
+                self.instance.condition.enterprise_customer_catalog_uuid = enterprise_customer_catalog_uuid
+                self.instance.condition.save()
+            else:
+                self.instance.condition = create_condition(
+                    EnterpriseCustomerCondition,
+                    enterprise_customer=enterprise_customer,
+                    enterprise_customer_catalog_uuid=enterprise_customer_catalog_uuid,
+                )
```
Comment (UmanShahzad) [inline ecommerce/enterprise/forms.py]: @douglashall This is where things crash, AFAIK. The condition expects `enterprise_customer_name`, `enterprise_customer_uuid`, and `enterprise_customer_catalog_uuid`, so we need to pass those in correctly. This reflects the old state of things before we decided to use `enterprise_customer_name` directly. The same of course goes for the `if` part of the branch -- there only `enterprise_customer_name` remains.

Comment (douglashall) [reply ecommerce/enterprise/forms.py]: Oh wait, I have fixed that already. Let me push.

File: ecommerce/extensions/offer/models.py [hunk 8]
```diff
@@ -324,6 +324,22 @@ def all_products(self):
 
 
 class Condition(AbstractCondition):
+    enterprise_customer_uuid = models.UUIDField(
+        null=True,
+        blank=True,
+        verbose_name=_('EnterpriseCustomer UUID')
+    )
+    # De-normalizing the EnterpriseCustomer name for optimization purposes.
+    enterprise_customer_name = models.UUIDField(
```
Comment (UmanShahzad) [inline ecommerce/extensions/offer/models.py]: ^^^ @douglashall It's being saved as a UUID

**Comment by douglashall** [comment]
@mjfrey Any chance you could review this? I appreciate that the diff is large and I have attempted to separate the work into logical commits. Let me know if there is anything I can do to make the review go smoother for you.

**Comment by douglashall** [comment]
@mjfrey I have opened a new PR containing just the first commit from this branch as you requested. The second two commits on this branch depend on that first commit.

File: ecommerce/invoice/migrations/0005_auto_20171003_1029.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.10.7 on 2017-10-03 14:29
+from __future__ import unicode_literals
+
+import django_extensions.db.fields
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
```
Comment (mjfrey) [inline ecommerce/invoice/migrations/0005_auto_20171003_1029.py]: What is the reason for this change?  Have we also tested how long this might take in production to run?

File: ecommerce/extensions/refund/migrations/0003_auto_20171003_1029.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.10.7 on 2017-10-03 14:29
+from __future__ import unicode_literals
+
+import django_extensions.db.fields
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
```
Comment (mjfrey) [inline ecommerce/extensions/refund/migrations/0003_auto_20171003_1029.py]: Same here.  Why this change.  This table is quite large in production.

File: ecommerce/invoice/migrations/0005_auto_20171003_1029.py (hunk 9 above)
Comment (douglashall) [reply ecommerce/invoice/migrations/0005_auto_20171003_1029.py]: This must be a previously missed migration that got picked up by running `makemigrations`. Have we been ignoring the missing migrations warning on purpose?

File: ecommerce/extensions/refund/migrations/0003_auto_20171003_1029.py (hunk 10 above)
Comment (UmanShahzad) [reply ecommerce/extensions/refund/migrations/0003_auto_20171003_1029.py]: I think they just came along when running `makemigrations`.

File: ecommerce/invoice/migrations/0005_auto_20171003_1029.py (hunk 9 above)
Comment (mjfrey) [reply ecommerce/invoice/migrations/0005_auto_20171003_1029.py]: I do not see missing migrations.  We should remove these.

Comment (douglashall) [reply ecommerce/invoice/migrations/0005_auto_20171003_1029.py]: @mjfrey Sounds good to me. I will do that now.

**Comment by douglashall** [comment]
@mjfrey How are we looking here? Do you think we'll be able to merge this today?

**Review by mjfrey — APPROVED** [review]
LGTM.  What is the plan to phase out the Coupon based offers and use this?  Are you planning on removing the old code?

**Comment by douglashall** [comment]
@mjfrey Thanks!

I think we still need to support the ability to associate an EnterpriseCustomer to a coupon, but there is followup work here to modify the coupon admin tool to create an Enterprise Offer. I will write up that story now and get it prioritized in an upcoming sprint.
